third_party/nvfuser/csrc/scheduler/normalization.cpp

```diff
@@ -1,35 +1,256 @@
+  int64_t reg_per_thread = getEstimatedRegisterUsage(ip.vect * ip.batch);
+  int64_t threads_per_sm = getThreadsPerSMGivenRegPerThread(reg_per_thread);
+  int64_t blocks_per_sm =
+      getBlocksPerSM(threads_per_sm, threads_per_block, dev_prop->warpSize);
+  ip.gdimy = blocks_per_sm * device_multiprocessor_count;
+  const int64_t outer_iter_min = 8;
+  const int64_t gdimy_max = scheduler_utils::roundUpToN(
+      ceilDiv(outer_dim_numel, outer_iter_min), device_multiprocessor_count);
+  while (ip.gdimy > gdimy_max && blocks_per_sm > 1) {
+    blocks_per_sm -= 1;
+    ip.gdimy = blocks_per_sm * device_multiprocessor_count;
+  }
+
+  // Step-3, set OuterParams Iteration dim: unroll_factor, bdimx, gdimy (already
+  // done)
+  op.unroll_factor = inner_dim_numel >= 4096 ? 4 : 2;
+  op.gdimy = ip.gdimy;
+  op.bdimx = scheduler_utils::roundUpPow2Or8(
+      ceilDiv(inner_dim_numel / op.unroll_factor, op.gdimy));
+
+  // Step-4, set OuterParams Reduction dim: bdimy.
+  op.bdimy = threads_per_block / op.bdimx;
+
+  // Step-5, special case, when inner_dim_numel <= 1024, bdimx is usually small
+  // after divide by vect and batch. In this case, bdimy is used to parallelize
+  // outer_dim instead of inner_dim. This pattern is named multi reductions per
+  // block (mrpb).
+  if (inner_dim_numel <= 1024) {
+    rparams->multiple_reds_per_blk = true;
+    rparams->tidx_for_outer_reduction = true;
+    constexpr int64_t threads_per_block_mrpb = 512;
+
+    // Step-1, InnerParams, Reduction dim: vect(reuse), batch, bdimx
+    ip.batch = 1;
+    ip.bdimx = inner_dim_numel / ip.vect;
+
+    // Step-2, InnerParams, Iteration dim: gdimy, bdimy (same as op.bdimy)
+    reg_per_thread = getEstimatedRegisterUsage(ip.vect * ip.batch);
+    threads_per_sm = getThreadsPerSMGivenRegPerThread(reg_per_thread);
+    blocks_per_sm = getBlocksPerSM(
+        threads_per_sm, threads_per_block_mrpb, dev_prop->warpSize);
+    ip.gdimy = blocks_per_sm * device_multiprocessor_count;
+
+    // Step-3, OuterParams, Iteration dim: unroll_factor, bdimy, gdimy(same as
+    // ip.gdimy)
+    op.gdimy = ip.gdimy;
+    op.unroll_factor = std::min(ip.vect, (int64_t)2);
+    op.bdimy = std::min(
+        ceilDiv(inner_dim_numel / op.unroll_factor, op.gdimy),
+        scheduler_utils::safeDiv(threads_per_block_mrpb, ip.bdimx));
+    ip.bdimy = op.bdimy;
+
+    // Step-4, OuterParams, Reduction dim: bdimx (same as ip.bdimx)
+    op.bdimx = ip.bdimx;
+
+    if (op.bdimx % dev_prop->warpSize == 0) {
+      rparams->pad_inner_reduction_to_warp = true;
+      rparams->pad_outer_reduction_to_warp = true;
+    }
+    rparams->block_dim_iter_dom = ParallelType::TIDy;
+  } else {
+    rparams->block_dim_inner_reduction_extra = ParallelType::TIDy;
+  }
+
+  rparams->persistent_kernel = true;
+  rparams->fastest_dim = true;
+  rparams->combined_inner_outer = true;
+  rparams->vectorization_factor_tmp_gmem_read =
+      std::min((int64_t)4, op.unroll_factor);
```

Comment:
Aside from the question on why limiting the factor to 4, why is this min computation done here rather than above? For example, at line 197, `op.unroll_factor` is used to calculate `op.bdimy`. So, shouldn't `op.bdimy`  be updated again to reflect the actual unroll factor?

Same for the min computation of `ip.vect`.

Reply:
> since our max unroll_factor is set to 4, we are safe to simplify it to:

Maybe then asserting it's `<= max_vectorization_factor`?

Reply:
> The min comparasion with ip.vect is needed, e.g. if dtype is half and vect = 8, we need to limit vectorization_factor_tmp_gmem_write to 4 since its dtype is float. 

I understand these adjustments are necessary. My concern is that at this point should have finalized the parameter values and should just copy them to `rparams`. Since the parameters have inter-dependencies, changing one parameter may require updating some other parameters, and it's unclear which parameters can be changed independently and which need to be updated. 

For ease of maintaining the code, let's separate parameter tuning/adjustment with setting `rparams` with the parameters.

And for `iop.inner_vect`, is there any other parameter that should be updated when it's adjusted?